### PR TITLE
Filter out the nulls after slicing the batches.

### PR DESCRIPTION
**Pull request by firestarman**
This PR is to fix the NPE in `GpuShuffleExchange` by filtering out the nulls after slicing the batches for `GpuRangePartitioner` and `GpuSinglePartitioning`. Just like what the `GpuHashPartitioning` does [here](https://github.com/NVIDIA/spark-rapids/blob/branch-21.06/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuHashPartitioning.scala#L89). 

The root cause of this NPE is the batch array returned from the method [sliceInternalGpuOrCpu ](https://github.com/NVIDIA/spark-rapids/blob/branch-21.06/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala#L98)may contain nulls that generated by the function ['sliceBatch'](https://github.com/NVIDIA/spark-rapids/blob/branch-21.06/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala#L35) for the corner case of empty batch (row number is 0). So even a single batch is expected, we still need to filter out the null.

Closes https://github.com/NVIDIA/spark-rapids/issues/2421 

Signed-off-by: Firestarman <firestarmanllc@gmail.com>

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala [hunk 1]
```diff
@@ -111,7 +111,7 @@ trait GpuPartitioning extends Partitioning with Arm {
         sliceInternalOnGpu(numRows, partitionIndexes, partitionColumns)
       } else {
         sliceInternalOnCpu(numRows, partitionIndexes, partitionColumns)
-      }
+      }.filter(_ != null)
```
Comment (wbo4958) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala:114]: is it safe to add this here, I checked some of them like (GpuSinglePartitioning,  GpuRangePartitioner) don't filter out the null? 

BTW, if it is safe, This PR should get rid of the other redundant filter

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala (hunk 1 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala:114]: > is it safe to add this here, I checked some of them like (GpuSinglePartitioning, GpuRangePartitioner) don't filter out the null?
> 

That's why I add the filtering here, to make sure it takes effect for all the callers.

Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala:114]: Removed the redundant filter

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala (hunk 1 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala]: Some tests failed, yes, it is not safe to do the filtering here. Updated

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
I marked the PR as draft, because I am not sure whether the linked issue is reproduceable on branch-21.06 (it was 0.6 before), which is the base branch of the PR.

I tried several times locally but can not get the NPE.

**Review by revans2 — APPROVED** [review]
nit: if it is simple, and only if it is simple, can we get a test that causes this to happen more regularly?

**Review by jlowe — APPROVED** [review]

**Comment by firestarman** [comment]
Thanks @viadea for the verification on branch-21.06. 
I am going to merge this PR since it works.

**Pull request merged**